### PR TITLE
ICU-22520 Replace char arrays with icu::CharString

### PR DESCRIPTION
**Pull request by roubert**
This also resolves oss-fuzz issue #65635 and ICU-22623.

##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22520
- [X] Required: The PR title must be prefixed with a JIRA Issue number.
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number.
- [X] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

ALLOW_MANY_COMMITS=true

### CODE REVIEW

File: icu4c/source/common/uresbund.cpp [hunk 1]
```diff
@@ -3161,31 +3165,31 @@ ures_getFunctionalEquivalent(char *result, int32_t resultCapacity,
         subStatus = U_ZERO_ERROR;
 
         if (res != nullptr) {
-            uprv_strcpy(found, ures_getLocaleByType(res, ULOC_VALID_LOCALE, &subStatus));
+            found.clear().append(ures_getLocaleByType(res, ULOC_VALID_LOCALE, &subStatus), subStatus);
         }
 
-        if (uprv_strcmp(found, parent) != 0) {
-            uprv_strcpy(parent, found);
+        if (found != parent.toStringPiece()) {
```
Comment (markusicu) [inline icu4c/source/common/uresbund.cpp:3171]: Can't we simply do
```
    if (found != parent) {
```
?

File: icu4c/source/common/uresbund.cpp [hunk 2]
```diff
@@ -3202,94 +3206,95 @@ ures_getFunctionalEquivalent(char *result, int32_t resultCapacity,
...
 
-        getParentForFunctionalEquivalent(found,res,&bund1,parent,1023);
+        getParentForFunctionalEquivalent(found.data(),res,&bund1,parent);
         ures_close(res);
         subStatus = U_ZERO_ERROR;
-    } while(!full[0] && *found && U_SUCCESS(*status));
+    } while(full.isEmpty() && !found.isEmpty() && U_SUCCESS(*status));
 
-    if((full[0]==0) && kwVal != defVal) {
+    if(full.isEmpty() && kwVal != defVal.toStringPiece()) {
```
Comment (markusicu) [inline icu4c/source/common/uresbund.cpp:3280]: direct comparison of two CharString's?

File: icu4c/source/common/uresbund.cpp [hunk 3]
```diff
@@ -3300,61 +3305,63 @@ ures_getFunctionalEquivalent(char *result, int32_t resultCapacity,
...
         } else if(omitDefault) {
 #if defined(URES_TREE_DEBUG)
-          fprintf(stderr,"Trim? full=%s, defLoc=%s, found=%s\n", full, defLoc, found);
+          fprintf(stderr,"Trim? full=%s, defLoc=%s, found=%s\n", full.data(), defLoc.data(), found.data());
 #endif        
-          if(uprv_strlen(defLoc) <= uprv_strlen(full)) {
+          if(defLoc.length() <= full.length()) {
             /* found the keyword in a *child* of where the default tag was present. */
-            if(kwVal == defVal) { /* if the requested kw is default, */
+            if(kwVal == defVal.toStringPiece()) { /* if the requested kw is default, */
```
Comment (markusicu) [inline icu4c/source/common/uresbund.cpp:3364]: ditto?

**Review by markusicu — APPROVED** [review]
nice!

**Review by FrankYFTang — APPROVED** [review]
This is a much better fix. Love it. Thanks. I review it very carefully. LGTM. Please land it ASAP

File: icu4c/source/common/uresbund.cpp (hunk 1 above)
Comment (roubert) [reply icu4c/source/common/uresbund.cpp:3171]: > Can't we simply do
> 
> ```
>     if (found != parent) {
> ```
> 
> ?

No, that is currently not possible, but I think that it should be, which is why I sent to PR #2764 for review last week to make that possible.

**Pull request merged**